Fix type information for `__version_info__` to handle prereleases

sirosen opened the pull request:
~Running `__version_info__ += __parsed_version__.pre` in `__init__.py` is incorrect (adding tuples). This bug will only be tripped on a prerelease version, so it hasn't had any impact to-date.~
For background, see also: https://github.com/marshmallow-code/webargs/issues/667
EDIT: The issue is not a bug at runtime, but a type annotation is needed for `__version_info__`. Right now, it combines strings and ints but has a type of `tuple[int, ...]`

`__version_info__` was being computed by adding together tuples. `mypy` has started flagging this (I think due to improved `packaging` type information).

To protect against flaws in the correction, and to present accurate type information, the `__version_info__` tuple is built by a utility function.

deckar01 commented:
> Unsupported operand types for + ("Tuple[int, ...]" and "Tuple[str, int]")

The current logic appears to be correct, the mypy error is just complaining that a string is being mixed in with integers. These tuples are intended to be compared regardless of length/type differences, so the type should probably be loosened to reflect that.

```py
__version_info__: tuple[int | str, ...] = __parsed_version__.release
```

sirosen commented:
> The current logic appears to be correct, the mypy error is just complaining that a string is being mixed in with integers.

You're 100% right. I almost never use `+` with tuples, so I got mixed up!
I thought this was citing lack of support for plus, but I was wrong.

> These tuples are intended to be compared regardless of length/type differences, so the type should probably be loosened to reflect that.

I know that in other repos we've used something akin to `MARSHMALLOW_MAJOR = marshmallow.__version_info__[0]` followed by checks like `if MARSHMALLOW_MAJOR < 3: ...`.
That makes me want to annotate this such that the first three elements are known to be ints.

Is what you suggest, `tuple[int | str, ...]` really preferable? Or is this better?
```python
__version_info__: tuple[int, int, int] | tuple[int, int, int, str, int]
```

(I think a helper is needed in this case, since the alternative is some pretty awkward use of `cast` in `__init__.py`)

lafrech commented:
> That makes me want to annotate this such that the first three elements are known to be ints.

Good point.

Would it be possible to just
- set type to `tuple[int, int, int] | tuple[int, int, int, str, int]`
- add some directive to skip the mypy check for the line since we (think we) know what we are doing, without going through the hassle of defining a helper
?

I can't think of a way to both set `type: ignore` and `type: tuple[int, int, int] | tuple[int, int, int, str, int]`.

If there's no simpler way, I'm fine with the helper. I'm just thinking this is a case where CI gets in the way.

sirosen committed:
Annotate the type of `__version_info__`

To avoid issues with `tuple[int, ...] + tuple[str, int]` on
pre-releases, which mypy flags when packaging data is available.

sirosen commented:
mypy is satisfied if we write
```
__version_info__: tuple[int, int, int] | tuple[int, int, int, str, int] = typing.cast(
    tuple[int, int, int], __parsed_version__.release
)
if __parsed_version__.pre:
    __version_info__ += __parsed_version__.pre  # type: ignore
```

The first line sets the type and the `+=` line doesn't change it.

I went ahead and amended to this version, since the helper is another thing to maintain and I'm not deeply committed to it.

`mypy` failed in CI as a result claiming that the `type: ignore` was unused. It was passing for me locally, so I'm going to try adding `packaging` to the requirements of the `mypy` pre-commit hook. I suspect it needs that type information and it's getting it for me from user site-packages or something similar.

sirosen committed:
Add `packaging` to mypy requirements

sirosen committed:
Fix version_info type annotation with ignore

Although the cast to `tuple[int, int, int]` is acceptable to `mypy`
and will work on newer pythons, on older pythons it raises an error at
runtime. The reason is that `from __future__ import annotations` does
not "enable" newer typing features in the sense that other future
imports change language behaviors. Rather, it enables one feature:
deferred evaluation of annotations.

In the cast, the type is evaluated at runtime (even though it is not
used). As a result, the deferred evaluation no longer applies, and we
must either use `typing.Tuple` or remove the cast.

As it is simpler to read, this changes the assignments in `__init__`
to simply `type: ignore[assignment]`.

sirosen commented:
I've had to tinker further, but this is now passing CI again. The sample I wrote above is acceptable to `mypy` and works on newer pythons, but fails at runtime under 3.7. It's surprising at first, but `cast(tuple[int, int, int], ...)` won't work under 3.7, even with the future import. My last commit message explains why in detail.

To workaround, I've switched to
```
__version_info__: tuple[int, int, int] | tuple[int, int, int, str, int] = (
    __parsed_version__.release  # type: ignore[assignment]
)
```
and a quick check of `reveal_type(__version_info__)` shows that this works as intended.

It also does appear to be necessary to include `packaging` for `mypy` to run consistently. As I suspected, that was the source of (spurious) warnings about unused ignore comments.

lafrech reviewed (approved):
Thanks for working on this.

This is exactly what I meant. Don't bother with a helper to construct the version, set the type to what it is and ignore mypy complaints.

lafrech commented:
See https://github.com/marshmallow-code/webargs/pull/680.

I think the mypy issue was an error in mypy that was fixed in mypy 0.931, although I couldn't find the changelog entry explaining what was fixed.

Still, this change makes the typing more specific without being too heavy (compared to the earlier version involving a helper method).

@sloria, @deckar01 any objection to this?

sloria reviewed (approved):
Looks good!
